src/Internal/ConcurrentIteratorChannel.php

```
@@ -71,24 +73,42 @@ public function onClose(\Closure $onClose): void
 
     public function receive(?Cancellation $cancellation = null): mixed
     {
-        if (!$this->receive->continue($cancellation)) {
-            $this->close();
-            throw new ChannelException("The channel closed while waiting to receive the next value");
+        try {
+            if ($this->receive->continue($cancellation)) {
+                return $this->receive->getValue();
+            }
+        } catch (DisposedException $exception) {
+            // Throw exception below
         }
 
-        return $this->receive->getValue();
+        throw new ChannelException(
+            "The channel closed while waiting to receive the next value",
+            0,
+            $exception ?? null,
+        );
     }
 
     public function send(mixed $data): void
     {
-        if ($data === null) {
-            throw new ChannelException("Cannot send null on a channel");
-        }
-
         if ($this->send->isComplete()) {
             throw new ChannelException("Cannot send on a closed channel");
         }
 
         $this->send->push($data);
     }
+
+    public function getIterator(): \Traversable
+    {
+        try {
+            foreach ($this->receive as $value) {
+                yield $value;
+            }
+        } catch (DisposedException $exception) {
+            throw new ChannelException(
+                "The channel closed while waiting to receive the next value",
+                0,
+                $exception,
+            );
```

Why does this still throw instead of simply returning?